New Airflow tutorial (#1033)
astronomer/docs: astro/cli/get-started.md

    ```

2. Open the directory:

    ```sh
    cd <your-astro-project-name>
    ```

3. Run the following Astro CLI command to initialize an Astro project in the directory:

    ```sh
    astro dev init
    ```

The Astro project is built to run Airflow with Docker. [Docker](https://docs.docker.com/) is a service to run software in virtualized containers within a machine. When you run Airflow on your machine with the Astro CLI, Docker creates a container for each Airflow component that is required to run DAGs. For this tutorial, you don't need an in-depth knowledge of Docker. All you need to know is that Airflow runs on the compute resources of your machine, and that all necessary files for running Airflow are included in your Astro project.

The default Astro project structure includes a collection of folders and files that you can use to run and customize Airflow. For this tutorial, you only need to know the following files and folders:

- `/dags`: A directory of DAG files. Each Astro project includes two example DAGs: `example-dag-basic` and `example-dag-advanced`. For more information on DAGs, see [Introduction to Airflow DAGs](https://www.astronomer.io/guides/dags/).

> - `Dockerfile`: This is where you specify your Astro Runtime, which a runtime environment that includes Apache Airflow and is built by Astronomer. The CLI generates new Astro projects with the latest version of Runtime, which is equivalent to the latest version of Airflow. For advanced use cases, you can also configure this file with Docker-based commands to run locally at build time.

Reviewer: ```suggestion
- `Dockerfile`: This is where you specify your version of Astro Runtime, which is a runtime environment that includes Apache Airflow and is built by Astronomer. The CLI generates new Astro projects with the latest version of Runtime, which is equivalent to the latest version of Airflow. For advanced use cases, you can also configure this file with Docker-based commands to run locally at build time.
```

## Step 2: Start Airflow

> Now that you have an Astro project ready, the next step is to actually start Airflow on your machine. In your terminal, open your Astro project directory and run the following command to start Airflow:

Reviewer: ```suggestion
Now that you have an Astro project ready, the next step is to actually start Airflow on your machine. In your terminal, open your Astro project directory and run the following command:
```

```sh
astro dev start
```

> Starting Airflow for the first time can take 2 to 5 minutes. Once your local environment is ready the CLI automatically opens your default web browser to the Airflow UI at `https://localhost:8080`.

Reviewer: ```suggestion
Starting Airflow for the first time can take 2 to 5 minutes. Once your local environment is ready, the CLI automatically opens a new tab or window in your default web browser to the Airflow UI at `https://localhost:8080`.
```

:::info

If port 8080 or 5432 are in use on your machine, Airflow won't be able to start. To run Airflow on alternative ports, run:

```sh
astro config set webserver.port <available-port>
astro config set postgres.port <available-port>
```

:::

## Step 3: Log in to the Airflow UI

> The [Airflow UI](https://www.astronomer.io/guides/airflow-ui/) is essential for managing Airflow. It contains information about your current DAG runs and configuration menus for Airflow connections, variables, and pools.

Reviewer: ```suggestion
The [Airflow UI](https://www.astronomer.io/guides/airflow-ui/) is essential for managing Airflow. It contains information about your current DAG runs and is the best place to create and update Airflow connections to third-party data services. For this step, let's just log in.
```

- We don't mention variables and pools anywhere so let's leave out
- Better define what actually happens in this step?

Author: I don't think we need the last sentence because of the step name

> To access the Airflow UI, open `http://localhost:8080/` in a browser and log in with `admin` for both your username and password.

Reviewer: Wouldn't be terrible to have a screenshot of this view

Author: Of the login? I think that might be a bit much, would rather have people focus on the next screenshot of the actual UI

The default page in the Airflow UI is the **DAGs** page, which shows an overview of all DAGs in your Airflow environment:

![View of starter DAGs in the Airflow UI](/img/docs/tutorial-airflow-ui.png)

> Each DAG is listed with its tags, owner, previous runs, schedule, timestamp of the last and next run, and the states of recent tasks. Because you haven't run any DAGs yet, the **Runs** and **Recent Tasks** sections are empty. Let's fix that!

Reviewer: ```suggestion
Each DAG is listed with a few of its properties, including tags, owner, previous runs, schedule, timestamp of the last and next run, and the states of recent tasks. Because you haven't run any DAGs yet, the **Runs** and **Recent Tasks** sections are empty. Let's fix that!
```

## Step 4: Trigger a DAG run

Let's run the `example-dag-basic` DAG that was generated with your Astro project. To provide a very basic demonstration of an ETL pipeline, this DAG creates an example JSON string, calculates a value based on the string, and prints the results of the calculation to the Airflow logs.

1. Before you can run any DAG in Airflow, you have to unpause it. To unpause `example-dag-basic`, click the slider button next to its name. Once you unpause it, the DAG starts to run on the schedule defined in its code.

    ![Pause DAG slider in the Airflow UI](/img/docs/tutorial-unpause-dag.png)

2. While all DAGs can run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger `example-dag-basic` by clicking the play button under the **Actions** column. During development, running DAGs on demand can help you identify and resolve issues.

After you press **Play**, the **Runs** and **Recent Tasks** sections for the DAG start to populate with data.

![DAG running in the Airflow UI](/img/docs/tutorial-run-dag.png)

> These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high level summaries of your runs that won't make much sense until you learn more about how Airflow works. To get a better picture of how your DAG is running, let's explore some other views in Airflow.

Reviewer: ```suggestion
These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high-level summaries of your runs that won't make much sense until you learn more about how Airflow works. To get a better picture of how your DAG is running, let's explore some other views in Airflow.
```

6. Define the dependencies between the two tasks using bitshift operators:

    ```python
    t1 >> t2
    ```

    This notation is unique to Airflow. It both defines and visualizes the dependencies between tasks. `t1` is **upstream** of `t2`. meaning that `t1` must finish before `t2` can start.

    There are multiple ways to define dependencies in Airflow, but this is one of the simplest. For more options on how to define task dependencies, see [Managing Dependencies in Apache Airflow](https://www.astronomer.io/guides/managing-dependencies).

7. Save your code.

## Step 7: Run the new DAG

To view your new DAG in the Airflow UI, enter `http://localhost:8080/` in your browser. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory. Existing files are parsed for changes every 30 seconds, while new files can be detected every 5 minutes.

When your new DAG appears in the Airflow UI, you can run it to test it.

1. Start the new DAG and trigger a run like you did in [Step 4](getting-started.md#step-4-trigger-a-dag-run).

> 2. Click the name of your new DAG and open the **Graph** view. After your DAG runs, there should be a dark green border around the tasks in the graph showing that your run was successful.

Reviewer: Screenshot here might also not be terrible!

## Step 8: View task logs

> When you tell Airflow to print something to the terminal, the output appears in Airflow's logs. This is an important feature for unit testing or otherwise troubleshooting DAGs locally.

Reviewer: ```suggestion
When you tell Airflow to print something to the terminal, the output appears in Airflow task logs. Task logs are an important feature for unit testing or otherwise troubleshooting DAGs. If a task in your DAG fails, task logs are the best place to investigate why.
```

---
sidebar_label: 'Get started'
title: 'Get started with Apache Airflow '
id: get-started
description: Install Apache Airflow and deploy your first local project with the Astro CLI.
---

Getting started with Apache Airflow locally is easy with the Astro CLI.

> This tutorial guides you through the process of using the Astro CLI to create an Astro project, run your project in a local Airflow environment, and write DAGs. We will also cover how to use the Astronomer Registry to discover example DAGs and Airflow providers.

Reviewer 2: ```suggestion
This tutorial guides you through the process of using the Astro CLI to create an Astro project, run your project in a local Airflow environment, and write DAGs. You'll also learn how to use the Astronomer Registry to discover example DAGs and Airflow providers.
```

> After completing this tutorial, you will know how to:

Reviewer 2: ```suggestion
After you complete this tutorial, you'll be able to:
```

- Run a local Airflow environment using the Astro CLI.
- Manage the Astro project directory.
- Write a new DAG.
- Navigate the Airflow UI.
- Use code from the Astronomer Registry.

## Time to complete

> This tutorial can be completed in 1 hour or less.

Reviewer 2: ```suggestion
This tutorial takes approximately 1 hour to complete.
```

## Assumed knowledge

> To get the most out of this tutorial, you should know:

Reviewer 2: ```suggestion
To complete this tutorial, you should already have a good understanding of:
```

- Basic Airflow concepts. See [Introduction to Apache Airflow](https://www.astronomer.io/guides/intro-to-airflow).
- Basic Python. See the [Python Documentation](https://docs.python.org/3/tutorial/index.html).

## Prerequisites

> To use this tutorial, you need:

Reviewer 2: ```suggestion
To use this tutorial, you need:
```
Remove this sentence

- A terminal that accepts bash commands. This is pre-installed on most operating systems.
- [Python 3](https://www.python.org/downloads/).
- [Docker Desktop](https://docs.docker.com/get-docker/) (v18.09 or higher).
- The [Astro CLI](cli/configure-cli.md).

> Astronomer also recommends having an integrated development environment (IDE) for Python development, such as [VSCode](https://code.visualstudio.com/).

Reviewer 2: Should this be a prerequisite?

Author: It's recommended but not required, which is why we kept it separate from the proper requirements.

Reviewer 2: @jwitz I would suggest that it is identified as a prerequisite if it's recommended. Also, an IDE seems critical to success. Your call.

Reviewer: I'd be ok with adding as a pre-rec

## Step 1: Create an Astro project

To run data pipelines on Astro, you first need to create an Astro project, which contains the set of files necessary to run Airflow locally.

> 1. Create a new, empty folder or directory on your computer called `my-astro-project`.

Reviewer 2: ```suggestion
1. Create a new, empty folder or directory on your computer named `my-astro-project`.
```

2. Run the initialization command from the terminal with the Astro CLI:

```sh
astro dev init
```

> The Astro project is built to run Airflow with Docker. [Docker](https://docs.docker.com/) is a service to run software in virtualized containers within a machine. When you run Airflow on your computer with the Astro CLI, Docker creates a container for each Airflow component that is required to run DAGs. For this tutorial, no in-depth knowledge of Docker is needed. All you need to know is that Airflow will run in a containerized environment on your machine, and that all necessary files for running these containers are included in the default Astro project.

Reviewer 2: ```suggestion
The Astro project is built to run Airflow with Docker. [Docker](https://docs.docker.com/) is a service to run software in virtualized containers within a machine. When you run Airflow on your computer with the Astro CLI, Docker creates a container for each Airflow component that is required to run DAGs. For this tutorial, you don't need an in-depth knowledge of Docker. All you need to know is that Airflow runs in a containerized environment on your machine, and that all necessary files for running these containers are included in the default Astro project.
```
You're using _machine_ and _computer_ interchangeably. Use one term consistently.

> The default Astro project structure includes a collection of folders and files that you can use to run and customize Airflow.  For this tutorial, you only need to know the following files and folders:

Reviewer 2: ```suggestion
The default Astro project structure includes a collection of folders and files that you can use to run and customize Airflow.  For this tutorial, only the following files and folders are required:
```

Author: The other files _are_ required, but you don't need to interact with them/ know about them yet. I think the original wording is more correct.

Reviewer 2: @jwitz how does a user _know_ a file or folder? The existing wording is incorrect.

- `/dags`: A directory of DAG files. Each Astro project includes two example DAGs. For more information on DAGs, see [Introduction to Airflow DAGs](https://www.astronomer.io/guides/dags/).

> - `Dockerfile`: This is where you specify your Airflow version. For advanced use cases, you can also configure this file with Docker-based commands to run locally at build time.

Reviewer: Say something like "when you create an Astro project, the latest available version of Airflow will be assumed"

## Step 2: Start Airflow

Now that you have an Astro project ready, the next step is to actually start Airflow on your machine. In your terminal, open your Astro project directory and run the following command to start Airflow:

```sh
astro dev start
```

> Starting Airflow for the first time can take 2 to 5 minutes. Once your local environment is ready, your terminal should show the following text:

Reviewer 2: ```suggestion
Starting Airflow for the first time can take 2 to 5 minutes. When your local environment is ready, the following text should appear in your terminal:
```

```text
Project is running! All components are now available.

Airflow Webserver: <http://localhost:8080>
Postgres Database: localhost:5432/postgres
The default Airflow UI credentials are: admin:admin
The default Postgres DB credentials are: postgres:postgres

```

:::info

If port 8080 or 5432 are in use on your machine, Airflow won't be able to start. To run Airflow on alternative ports, run:

```sh
astro config set webserver.port <available-port>
astro config set postgres.port <available-port>
```

:::

## Step 3: Log in to the Airflow UI

The [Airflow UI](https://www.astronomer.io/guides/airflow-ui/) is essential for managing Airflow. It contains information about your current DAG runs and configuration menus for Airflow connections, variables, and pools.

> To access the Airflow UI, open `http://localhost:8080/` in a browser and log in with `admin` for both your username and password.

Author: I changed the text in the previous step, but I think it's okay to be a bit redundant here with the information about the URL/ login. The CLI behavior isn't necessarily relevant to the UI itself, which is what this step is all about.

The default page in the Airflow UI is the **DAGs** page, which shows an overview of all DAGs in your Airflow environment:

![View of starter DAGs in the Airflow UI](/img/docs/tutorial-airflow-ui.png)

Each DAG is listed with its tags, owner, previous runs, schedule, timestamp of the last and next run, and the states of recent tasks. Because you haven't run any DAGs yet, the **Runs** and **Recent Tasks** sections are empty. Let's fix that!

## Step 4: Trigger a DAG run

> Let's run the `example-dag-basic` DAG that was generated with your Astro project.

Author: I think that's clear enough, and potentially jumping the gun on explaining how to write their own DAGs

> 1. Before you can run a DAG, you have to unpause it. To unpause `example-dag-basic`, click on the slider button next to its name. Once you unpause it, the DAG will start to run on the schedule defined in its code:

Reviewer 2: ```suggestion
1.  Click the slider next to `example-dag-basic` to start the DAG. The DAC runs on the schedule defined in its code.
```

> 1. Before you can run a DAG, you have to unpause it. To unpause `example-dag-basic`, click on the slider button next to its name. Once you unpause it, the DAG will start to run on the schedule defined in its code:

Author: Again, might be too soon to drop this info. To me, this is more important for the DAG writing step - right now, we just want to get the DAG going.

    ![Pause DAG slider in the Airflow UI](/img/docs/tutorial-unpause-dag.png)

> 2. While all DAGs can run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger `example-dag-basic` by clicking the play button under the **Actions** column. During development, manually triggering DAG runs can be helpful with debugging.

Reviewer 2: ```suggestion
2. Click **Play** in the **Actions** column of the `example-dag-basic` DAG  to run it on demand. During development, running DAGs on demand can help you identify and resolve issues.
```

After you press play, the **Runs** and **Recent Tasks** sections for the DAG start to populate with information in each circle.

![DAG running in the Airflow UI](/img/docs/tutorial-run-dag.png)

> These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high level summaries of your runs that won't make much sense until you learn more about how Airflow works. To get a better picture of how our DAG is running, let's explore some other views in Airflow.

Reviewer 2: ```suggestion
These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high level summaries of your runs that won't make much sense until you learn more about how Airflow works. To get a better picture of how your DAG is running, let's explore some other views in Airflow.
```

## Step 5: Explore the Airflow UI

The Airflow UI's navigation bar contains 5 tabs, each with different information about your Airflow environment. For more information about what you can find in each tab, see [The Airflow UI](https://www.astronomer.io/guides/airflow-ui/).

> For now, let's explore the available views in the **DAGs** page. To access different DAG views for `example-dag-basic`:

Reviewer 2: ```suggestion
Let's explore the available views in the **DAGs** page. To access different DAG views for `example-dag-basic`:
```

> 1. Click on the name of the DAG.

Reviewer 2: ```suggestion
1. Click the DAG name.
```
Don't use _Click **on**_. Use Click. Correct all instances.

    The default DAG view is the **Grid** view, which shows the state of completed and currently running tasks. Each column in the grid represents a complete DAG run, and each block in the column represents a specific task instance.

    ![Grid view](/img/docs/tutorial-grid-view.png)

>     Clicking on a square will open additional information about the related task instance on the righthand side of the UI. The task instance view includes tabs that include additional information for the task instance, such as its logs and historic runs. This is one of many available views that show details about your DAG. For now, let's explore two more views.

Reviewer 2: ```suggestion
    Click on a square to display additional information about the related task instance on the right side of the Airflow UI. The task instance view includes tabs with additional information for the task instance, such as its logs and historic runs. This is one of many available views that show details about your DAG. Let's explore two more views.
```

> 2. In the tab with **Grid** highlighted, click **Graph**. This view shows the dependencies between your tasks and is useful for troubleshooting issues that stem from dependencies. You can determine the state of your current task runs by checking the colors of the borders around teach task.

Reviewer 2: ```suggestion
2. In the **Grid** tab, click **Graph**. This view shows task dependencies and relationships and can help you troubleshoot dependency issues. The border colors of each task indicate the task run state.
```

    ![Graph view](/img/docs/tutorial-graph-view.png)

> 3. In the tab with **Graph** highlighted, click **Code**. This view shows the source code for your DAG.

Reviewer 2: ```suggestion
3. In the **Graph** tab, click **Code** to display your DAG source code.
```

    ![Code view](/img/docs/tutorial-code-view.png)

  :::info

>   While DAG code can be viewed from within the Airflow UI, code edits have to be done directly in the Python file within the `/dags` folder. The displayed code will every 30 seconds.

Reviewer 2: ```suggestion
  While you can view DAG code within the Airflow UI, code edits must be completed in the Python file within the `/dags` folder. The displayed code updates every 30 seconds.
```

>   While DAG code can be viewed from within the Airflow UI, code edits have to be done directly in the Python file within the `/dags` folder. The displayed code will every 30 seconds.

Reviewer 3: ```suggestion
  While DAG code can be viewed from within the Airflow UI, code edits have to be done directly in the Python file within the `/dags` folder. The displayed code will update every 30 seconds.
```

  :::

## Step 6: Write a new DAG

> In this step, we'll write a DAG that multiplies an input by 23. You will copy most of the code, trigger the DAG, and check to make sure you get the expected output.

Reviewer 2: ```suggestion
In this step, you'll write a DAG that multiplies an input by 23. You'll copy most of the code, trigger the DAG, and then confirm the expected output is returned.
```

> 1. Create a new Python file in the `/dags` folder named `my-dag.py`

Reviewer 2: ```suggestion
1. Create a new Python file in the `/dags` folder named `my-dag.py`.
```

> 2. Open `my-dag.py` in your IDE. Start by adding the required imports for Python packages:

Reviewer 2: ```suggestion
2. Open `my-dag.py` in your IDE. Add the required imports for Python packages:
```

    ```python
    from airflow import DAG
    from airflow.operators.python import PythonOperator
    from airflow.operators.bash import BashOperator

    from datetime import datetime, timedelta

    ```

>     The first line imports the `DAG` class, the second and third line import two Airflow operators that we use in this example, and the last line imports two objects from the [datetime package](https://docs.python.org/3/library/datetime.html, which is required to define the schedule of a DAG.

Reviewer 2: ```suggestion
    The first line imports the `DAG` class, the second and third lines import two Airflow operators that are used in this example, and the last line imports two objects from the [datetime package](https://docs.python.org/3/library/datetime.html), which are required to define the schedule of a DAG.
```

>     The first line imports the `DAG` class, the second and third line import two Airflow operators that we use in this example, and the last line imports two objects from the [datetime package](https://docs.python.org/3/library/datetime.html, which is required to define the schedule of a DAG.

Reviewer 3: ```suggestion
    The first line imports the `DAG` class, the second and third line import two Airflow operators that we use in this example, and the last line imports two objects from the [datetime package](https://docs.python.org/3/library/datetime.html), which is required to define the schedule of a DAG.
```

> 3. In the same file, add two static variables and a simple Python function that multiplies an input by 23. We will call this code later in our DAG using an Airflow operator.

Reviewer 2: ```suggestion
3. Add two static variables and a Python function that multiplies an input by 23. Your DAG uses an Airflow operator to call this code.
```

    ```python
    # constants
    MY_NAME = "MY_NAME"
    MY_NUMBER = 19

    def multiply_by_23(number):
        """Multiplies a number by 23 and prints the results to Airflow logs."""
        result = number * 23
        print(result)

    ```

4. Add a `dag` object, which is an instance of the `DAG` class:

    ```python
    with DAG(
        dag_id="my_first_dag",
        start_date=datetime(2022,7,28),
        schedule_interval=timedelta(minutes=30),
        catchup=False,
        tags= ["tutorial"],
        default_args={
            "owner": MY_NAME,
            "retries": 2,
            "retry_delay": timedelta(minutes=5)
        }
    ) as dag:

    ```

    `with DAG(...) as dag:` instantiates a DAG context in which tasks can be defined and given depenencies. The instantiation includes several important arguments:

>     - `dag_id` (Required): The name of the DAG that appears in the Airflow UI. Each DAG must have unique name, and it's best practice to use the same name for both the DAG file and the `dag_id`.

Reviewer 2: ```suggestion
    - `dag_id` (Required): The name of the DAG that appears in the Airflow UI. Each DAG must have unique name, and Astronomer recommends using the same name for the DAG file and the `dag_id`.
```

>     - `start_date` (Required): The date and time from which the DAG will be scheduled, given as a datetime object. In this example, the DAG will be triggered on its schedule as long as the current time is 0:00 UTC on July 28th, 2022 or later.

Reviewer 2: ```suggestion
    - `start_date` (Required): The date and time from the DAG is scheduled to run, given as a datetime object. In this example, the DAG is triggered on its schedule as long as the current time is 0:00 UTC on July 28th, 2022 or later.
```

>     - `schedule_interval`: The frequency at which this DAG runs. You can define this as a timedelta object, a [CRON expression](https://crontab.guru/), or a macro like "@daily". If you don't set this value, the DAG runs every 24 hours after the `start_date`.

Reviewer 2: ```suggestion
    - `schedule_interval`: The frequency the DAG runs. You can define this as a timedelta object, a [CRON expression](https://crontab.guru/), or as a macro such as "@daily". If you don't set this value, the DAG runs every 24 hours after the `start_date`.
```

>     - `catchup`: Determines whether the DAG reruns all DAG runs that were scheduled before today's date. The default value is `True`, but it's best practice to set this argument to `False` unless you are explicitly running your DAG to backfill runs.

Author: "Best practice" here references the overall consensus of the Airflow community vs. just Astronomer's recommendation. I think the original wording is important to preserve in this instance

>     - `catchup`: Determines whether the DAG reruns all DAG runs that were scheduled before today's date. The default value is `True`, but it's best practice to set this argument to `False` unless you are explicitly running your DAG to backfill runs.

Reviewer 2: @jwitz use  _...but it is recommended that you set..._ instead.

    - `tags`: Defines the **Tags** that appear in the **DAGs** page of the Airflow UI. These can help you organize DAGs in more complex projects.

>     - `default_args`: A list of configurations for the DAG's behavior. In this DAG, we use these arguments to change the owner of the DAG, give the DAG a maximum of two retries in case it fails, and tell the DAG to wait five minutes before retrying. Many more arguments can be passed to a DAG at instantiation. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dag/index.html) for a complete list.

Reviewer 2: ```suggestion
    - `default_args`: A list of configurations for the DAG's behavior. In this DAG, these arguments change the owner of the DAG, give the DAG a maximum of two retries in case of failure, and tell the DAG to wait five minutes before retrying. Many more arguments can be passed to a DAG at instantiation. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dag/index.html) for a complete list.
```

4. Add an operator to your DAG:

    ```python
    t1 = BashOperator(
        task_id="say_my_name",
        bash_command=f"echo {MY_NAME}"
    )

> 

Reviewer 3: ```suggestion
```

    ```

>     An operator is a Python class containing the logic to define the work to be completed by a single task. Our first task (`t1`) uses the [BashOperator](https://registry.astronomer.io/providers/apache-airflow/modules/bashoperator) to run a bash command that prints the `MY_NAME` variable to the terminal. The `task_id` defines the name of the task that appears in the Airflow UI. Each task requires a unique `task_id`.

Reviewer 2: ```suggestion
    An operator is a Python class containing the logic to define the work to be completed by a single task. The first task (`t1`) uses the [BashOperator](https://registry.astronomer.io/providers/apache-airflow/modules/bashoperator) to run a bash command that prints the `MY_NAME` variable to the terminal. The second task  (`task_id`) defines the name of the task that appears in the Airflow UI. Each task requires a unique `task_id`.
```

5. Add a second operator to your DAG:

    ```python
    t2 = PythonOperator(
        task_id="multiply_my_number_by_23",
        python_callable=multiply_by_23,
        op_kwargs={"number": MY_NUMBER}
    )
    ```

>     Our second task (`t2`) uses the [PythonOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator) to call the `mulitply_by_23` function that we defined earlier. We can then pass arguments to the Python function by defining a dictionary of key value pairs in the `op_kwargs` parameter.

Reviewer 2: ```suggestion
    The second task (`t2`) uses the [PythonOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator) to call the `mulitply_by_23` function that you defined earlier. To pass arguments to the Python function, you define a dictionary of key value pairs in the `op_kwargs` parameter.
```

>     Our second task (`t2`) uses the [PythonOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator) to call the `mulitply_by_23` function that we defined earlier. We can then pass arguments to the Python function by defining a dictionary of key value pairs in the `op_kwargs` parameter.

Reviewer 3: ```suggestion
    Our second task (`t2`) uses the [PythonOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator) to call the `multiply_by_23` function that we defined earlier. We can then pass arguments to the Python function by defining a dictionary of key value pairs in the `op_kwargs` parameter.
```

Reviewer: Half a sentence explanation of what the PythonOperator does?

    This is one of the features that makes Airflow so powerful: Any action that can be defined in Python, no matter how complex, can be orchestrated using Airflow.

6. Define the dependencies between the two tasks using bit-shift operators:

    ```python
    t1 >> t2
    ```

    This notation is unique to Airflow. It both defines and visualizes the dependencies between tasks. `t1` is **upstream** of `t2`. meaning that `t1` must finish before `t2` can start.

    There are multiple ways to define dependencies in Airflow, but this is one of the simplest. For more options on how to define task dependencies, see [Managing Dependencies in Apache Airflow](https://www.astronomer.io/guides/managing-dependencies).

7. Save your code.

## Step 7: Run the new DAG

> To see your new DAG in the Airflow UI, run your browser at `http://localhost:8080/`. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory. Existing files will be parsed for changes every 30 seconds while new files can be detected every 5 minutes.

Reviewer 2: ```suggestion
To view your new DAG in the Airflow UI, enter `http://localhost:8080/` in your browser. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory. Existing files are parsed for changes every 30 seconds, while new files can be detected every 5 minutes.
```

> To see your new DAG in the Airflow UI, run your browser at `http://localhost:8080/`. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory. Existing files will be parsed for changes every 30 seconds while new files can be detected every 5 minutes.

Reviewer: ```suggestion
To see your new DAG in the Airflow UI, refresh your browser at `http://localhost:8080/`. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory. Existing files will be parsed for changes every 30 seconds while new files can be detected every 5 minutes.
```

> Once your new DAG appears in the Airflow UI, we can run it:

Reviewer 2: ```suggestion
When your new DAG appears in the Airflow UI, you can run it to test it.
```

> 1. Unpause the new DAG and trigger a run like you did in Step 4.

Reviewer 2: ```suggestion
1. Start the new DAG and trigger a run. See <link>.
```

> 2. Click the name of your new DAG and open **Graph** view. After your DAG runs, there should be a dark green border around the tasks in the graph showing that your run was successful.

Reviewer 3: ```suggestion
2. Click the name of your new DAG and open the **Graph** view. After your DAG runs, there should be a dark green border around the tasks in the graph showing that your run was successful.
```

## Step 8: View task logs

> When we tell Airflow to print something to the terminal, the output appears in Airflow's logs. To see the output of our tasks:

Reviewer 2: ```suggestion
When you tell Airflow to print something to the terminal, the output appears in Airflow's logs.
```
Why is this important? Why is this helpful?

1. In the Airflow UI, open either the **Grid** or the **Graph** view.  
2. Click the `say_my_name` task to make the task instance view appear.

> 3. Click the **Log** button.

Reviewer 2: ```suggestion
3. Click **Log**.
```
Avoid references to UI elements (button).

> In the log output, you should see the string we set for the `MY_NAME` constant:

Reviewer 2: ```suggestion
In the log output, you should see the string you set for the `MY_NAME` constant:
```

```text
[2022-07-29, 15:17:19 UTC] {subprocess.py:74} INFO - Running command: ['bash', '-c', 'echo MY_NAME']
[2022-07-29, 15:17:19 UTC] {subprocess.py:85} INFO - Output:
[2022-07-29, 15:17:19 UTC] {subprocess.py:92} INFO - MY_NAME
[2022-07-29, 15:17:19 UTC] {subprocess.py:96} INFO - Command exited with return code 0
[2022-07-29, 15:17:19 UTC] {taskinstance.py:1415} INFO - Marking task as SUCCESS. dag_id=my_first_dag, task_id=say_my_name, execution_date=20220729T144715, start_date=20220729T151718, end_date=20220729T151719

```

> Follow the same steps for the `multiply_my_number_by_23` task. The task logs should include the results of our multiplication operation.

Reviewer 2: Repeat what steps? Be explicit.

```text
[2022-07-29, 15:17:22 UTC] {logging_mixin.py:115} INFO - 437
[2022-07-29, 15:17:22 UTC] {python.py:173} INFO - Done. Returned value was: None
[2022-07-29, 15:17:22 UTC] {taskinstance.py:1415} INFO - Marking task as SUCCESS. dag_id=my_first_dag, task_id=multiply_my_number_by_23, execution_date=20220729T144715, start_date=20220729T151721, end_date=20220729T151722
```

## Step 9: Use the Astronomer Registry

The example DAG you wrote used two different core Airflow operators, but there are many more operators available outside of core Airflow. The best place to explore and search for operators is the [Astronomer Registry](https://registry.astronomer.io/).

> The Astronomer Registry is also home to a growing set of [example DAGs](https://registry.astronomer.io/dags) that demonstrate more complex use cases and integrations. Let's try adding an example DAG to our Astro project:

Reviewer 2: ```suggestion
You can find [example DAGs](https://registry.astronomer.io/dags) in the Astronomer Registry that demonstrate more complex use cases and integrations. Now you'll add an example DAG to the Astro project:
```

1. Go to the [Astronomer Registry](https://registry.astronomer.io/).
2. Click **Browse DAGs**.
3. In the search bar, search for `TaskFlow API`.
4. Open the **TaskFlow API ETL Example** DAG.
5. In the DAG's information page, click the **Code** tab and copy the contents of the DAG code.
6. Paste the code into a new `.py` in the `dags` folder of your Astro project.
7. Run the DAG from the Airflow UI.

> This DAG demonstrates how you can define task dependencies using the TaskFlow API, which is a popular, more Pythonic method of writing DAGs. Instead of defining dependencies with bitshift operators, we can do so with direct calls to functions that have been decorated as tasks. In the example DAG, the following line:

Reviewer 2: ```suggestion
This DAG demonstrates how you can define task dependencies using the TaskFlow API, which is a popular, more Pythonic method of writing DAGs. Instead of defining dependencies with bitshift operators, you can use direct calls to functions that have been decorated as tasks. In the example DAG, the following line:
```

> This DAG demonstrates how you can define task dependencies using the TaskFlow API, which is a popular, more Pythonic method of writing DAGs. Instead of defining dependencies with bitshift operators, we can do so with direct calls to functions that have been decorated as tasks. In the example DAG, the following line:

Reviewer 3: Do we use "bit-shift operator" (line 232) or "bitshift operator"? I couldn't find anything in the style guide and googling I get most results for "bit shift operator" without a dash 😅

---
sidebar_label: 'Get started'
title: 'Get started with Apache Airflow '
id: get-started

> description: Deploy your first local Apache Airflow project with the Astro CLI.

Reviewer: ```suggestion
description: Install Apache Airflow and deploy your first local project with the Astro CLI.
```

Can we keep the word `install` in here? I know we don't typically use it, but I wanna make sure this comes up when someone googles `install airflow`

---

Getting started with Apache Airflow locally is made easy thanks to the Astro CLI.

This tutorial guides you through the process of using the Astro CLI to create an Astro project, run your project in a local Airflow environment, and write DAGs. We will also cover how to use the Astronomer Registry to discover example DAGs and Airflow providers.

## Time to complete

This tutorial can be completed in 1 hour or less.

## Assumed knowledge

To get the most out of this tutorial, you should know:

- Basic Airflow concepts. See [Introduction to Apache Airflow](https://www.astronomer.io/guides/intro-to-airflow).
- Basic Python. See the [Python Documentation](https://docs.python.org/3/tutorial/index.html).

## Prerequisites

To use this tutorial, you need:

> - A terminal that accepts bash commands (pre-installed on most operating systems).

Reviewer: ```suggestion
- A terminal that accepts bash commands. This is pre-installed on most operating systems.
```

- [Python 3](https://www.python.org/downloads/).
- [Docker Desktop](https://docs.docker.com/get-docker/) (v18.09 or higher).
- The [Astro CLI](cli/configure-cli.md).

Astronomer also recommends having an integrated development environment (IDE) for Python development, such as [VSCode](https://code.visualstudio.com/).

## Step 1: Create an Astro project

> To run Airflow pipelines on Astro, you first need to create an Astro project, which contains the set of files necessary to run Airflow.
> 
> After [installing the Astro CLI](cli/configure-cli.md), create a folder for the new Astro project and run the initialization command from the terminal:

Reviewer: ```suggestion
To run data pipelines on Astro, you first need to create an Astro project, which contains the set of files necessary to run Airflow locally.

1. Create a new, empty folder or directory on your computer called `my-astro-project`.
2. Run the initialization command from the terminal with the Astro CLI:
```

- We already had the CLI installed as pre-requisite, so no need to link again?
- Let's make this steps and guided as much as poss

```sh
astro dev init
```

> The default Astro project is built to run Airflow in a Dockerized environment. [Docker](https://docs.docker.com/) is a service to run software in virtualized containers within a machine. When you run Airflow locally on your local machine the Astro CLI, Docker creates a container for each Airflow component that is required to run DAGs. For this tutorial, no in-depth knowledge of Docker is needed. All you need to know is that Airflow will run in a containerized environment on your machine, and all necessary files for running these containers are included in the default Astro project.

Reviewer: ```suggestion
The Astro project is built to run Airflow with Docker. [Docker](https://docs.docker.com/) is a service to run software in virtualized containers within a machine. When you run Airflow on your computer with the Astro CLI, Docker creates a container for each Airflow component that is required to run DAGs. For this tutorial, no in-depth knowledge of Docker is needed. All you need to know is that Airflow will run in a containerized environment on your machine, and that all necessary files for running these containers are included in the default Astro project.
```

I wonder if we could cut more here but maybe not

> The default Astro project includes a collection of folders and files that you can use to run and Customize Airflow. For this tutorial, you only need to understand the following files and folders:

Reviewer: ```suggestion
For this tutorial, you only need to understand the following files that exist in every Astro project:
```

- `/dags`: A directory of DAG files. Each Astro project includes two example DAGs. For more information on DAGs, see [Introduction to Airflow DAGs](https://www.astronomer.io/guides/dags/).

> - `/tests` : A directory for unit tests that you can run with `astro dev pytest`. The default Astro project includes a few basic unit tests.

Reviewer: ```suggestion
- `/tests` : A directory for unit tests that you can run locally. They help make sure that code is tested before it's promoted to production. The default Astro project includes a few basic unit tests, but your team could customize them and create new ones as you scale.
```

- We haven't said anything about `astro dev parse`
- I think let's be direct about what each of these can do

Author: Whoops, I'm not sure why this is still included in the list of files. because it's not touched at all during the tutorial, I think we should remove it

> - `Dockerfile`: A list of commands that Docker runs when building containers for your Airflow environment.- `Dockerfile`: A list of commands that Docker runs when building containers for your Airflow environment. This is where your Airflow version and distribution are specified.

Reviewer: ```suggestion
- `Dockerfile`: This is where you specify your Airflow version. For advanced use cases, you can also configure this file with Docker-based commands to run locally at build time.
```

- `packages.txt`: A list of OS-level packages that Docker installs when building your Airflow environment.

> - `requirements.txt`: A list of Python-level packages that Docker installs when building your Airflow environment.

Reviewer: Could list 1-2 common examples of packages for each

Author: Because we don't touch this in the tutorials anymore, should we remove it?

## Step 2: Start Airflow

> In your terminal, open your Astro project directory and run the following command to start Airflow:

Reviewer: ```suggestion
Now that you have an Astro project ready, the next step is to actually start Airflow on your machine. In your terminal, open your Astro project directory and run the following command to start Airflow:
```

```sh
astro dev start
```

> Starting Airflow for the first time can take 2-5 minutes. Once the CLI successfully starts Airflow, your terminal should show the following text:

Reviewer: ```suggestion
Starting Airflow for the first time can take 2 to 5 minutes. Once your local environment is ready, your terminal should show the following text:
```

```
Project is running! All components are now available.

Airflow Webserver: <http://localhost:8080>
Postgres Database: localhost:5432/postgres
The default Airflow UI credentials are: admin:admin
The default Postgres DB credentials are: postgres:postgres

```

:::info

If port 8080 or 5432 are in use on your machine, Airflow won't be able to start. To run Airflow on alternative ports, run:

```sh
astro config set webserver.port <available-port>
astro config set postgres.port <available-port>
```

:::

## Step 3: Log in to the Airflow UI

> The [Airflow UI](https://www.astronomer.io/guides/airflow-ui/) is essential for managing Airflow. It contains information about your current DAG runs and configuration menus for Airflow connections, variables, and pools.

Reviewer: ```suggestion
One of the key components of a local Airflow environment is the [Airflow UI](https://www.astronomer.io/guides/airflow-ui/), which is essential for managing Airflow. It contains information about your current DAG runs and configuration menus for Airflow connections, variables, and pools.
```

Author: First sentence seems redundant with it itself now

To access the Airflow UI, open `http://localhost:8080/` in a browser and log in with `admin` for both your username and password.

The default page in the Airflow UI is the **DAGs** page, which shows an overview of all DAGs in your Airflow environment:

![DAGView.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/ec24bc62-a4fb-4aa2-a1d0-fa0f9521d296/DAGView.png)

Each DAG is listed with its tags, owner, previous runs, schedule, timestamp of the last and next run, and the states of recent tasks. Because you haven't run any DAGs yet, the **Runs** and **Recent Tasks** sections are empty. Let's fix that!

## Step 4: Trigger a DAG run

Let's run the `example-dag-basic` DAG that was generated with your Astro project.

> 1. Before you can run a DAG, you have to unpause it. To unpause `example-dag-basic`, click on the slider button next to its name. Once you unpause it, the DAG runs once for its latest scheduled date.

Reviewer: ```suggestion
1. Before you can run a DAG, you have to unpause it. To unpause `example-dag-basic`, click on the slider button next to its name. Once you unpause it, the DAG runs once.
```

Author: I think this context is important for understanding _why_ it runs once.

2. While all DAGs can run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger `example-dag-basic` by clicking the play button under the **Actions** column. During development, manually triggering DAG runs can be helpful with debugging.

After you press play, the **Runs** and **Recent Tasks** sections for the DAG start to populate with information. These are only high level summaries of your runs. Next, let's explore the other ways to view DAGs in Airflow.

## Step 5: Explore the Airflow UI

The Airflow UI's navigation bar contains 5 tabs, each with different information about your Airflow environment. For more information about what you can find in each tab, see [The Airflow UI](https://www.astronomer.io/guides/airflow-ui/).

For now, let's explore the available views in the **DAGs** page. To access different DAG views for `example-dag-basic`, click on its name.

> The default DAG view is the **Grid** view, which shows the state of completed and currently running tasks. Each column in the grid represents a complete DAG run, and each block in the column represents a specific task instance. Clicking on a square will open additional information about the related task instance on the righthand side of the UI. The task instance view includes tabs that include additional information for the task instance, such as its logs and historic runs.

Reviewer: Steps? Have them click into tree view (list benefit) then Grid view (list benefit)?

The **Grid** view is one of many available views that show details about your DAG. For now, let's explore one more view. In the tab with **Grid** highlighted, click **Code**. This view shows the source code for your DAG.

:::info

While DAG code can be viewed from within the Airflow UI, code edits have to be done directly in the Python file within the `/dags` folder. The displayed code will every 30 seconds.

:::

## Step 6: Write a new DAG

> Now that you know how to start an Airflow environment and run DAGs, you can start writing and testing your own DAGs.

Reviewer: Better intro to this step? Are we gonna write a DAG that does something, or are we copy pasting an example?

> "In this step, we're going to write a DAG that multiplies an input by 23. You will copy most of the code, trigger the DAG, and see if you get an expected output"

Something like that?

1. Create a new Python file in the `/dags` folder named `my-dag.py`

> 2. Open the file in your IDE. Start by adding the required imports for Python packages:

Reviewer: ```suggestion
2. Open `my-dag.py` in your IDE. Start by adding the required imports for Python packages:
```

        python_callable=multiply_by_23,
        op_kwargs={"number": MY_NUMBER}
    )
    ```

    Our second task (`t2`) uses the [PythonOperator](https://registry.astronomer.io/providers/apache-airflow/modules/pythonoperator) to call the `mulitply_by_23` function that we defined earlier. We can then pass arguments to the Python function by defining a dictionary of key value pairs in the `op_kwargs` parameter.

    This is one of the features that makes Airflow so powerful: Any action that can be defined in Python, no matter how complex, can be orchestrated using Airflow.

6. Define the dependencies between the two tasks using bit-shift operators:

    ```python
    t1 >> t2
    ```

    This notation is unique to Airflow. It both defines and visualizes the dependencies between tasks. `t1` is **upstream** of `t2`. meaning that `t1` must finish before `t2` can start.

    There are multiple ways to define dependencies in Airflow, but this is one of the simplest. For more options on how to define task dependencies, see [Managing Dependencies in Apache Airflow](https://www.astronomer.io/guides/managing-dependencies).

7. Save your code.

## Step 7: Run the new DAG

> Refresh your browser at `http://localhost:8080/`. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory. Existing files will be parsed for changes every 30 seconds while new files can be detected every 5 minutes.

Reviewer: Better intro to this step as well? Add steps too?

In the log output, you should see the string we set for the `MY_NAME` constant:

```text
[2022-07-29, 15:17:19 UTC] {subprocess.py:74} INFO - Running command: ['bash', '-c', 'echo MY_NAME']
[2022-07-29, 15:17:19 UTC] {subprocess.py:85} INFO - Output:
[2022-07-29, 15:17:19 UTC] {subprocess.py:92} INFO - MY_NAME
[2022-07-29, 15:17:19 UTC] {subprocess.py:96} INFO - Command exited with return code 0
[2022-07-29, 15:17:19 UTC] {taskinstance.py:1415} INFO - Marking task as SUCCESS. dag_id=my_first_dag, task_id=say_my_name, execution_date=20220729T144715, start_date=20220729T151718, end_date=20220729T151719

```

Follow the same steps for the `multiply_my_number_by_23` task. The task logs should include the results of our multiplication operation.

```text
[2022-07-29, 15:17:22 UTC] {logging_mixin.py:115} INFO - 437
[2022-07-29, 15:17:22 UTC] {python.py:173} INFO - Done. Returned value was: None
[2022-07-29, 15:17:22 UTC] {taskinstance.py:1415} INFO - Marking task as SUCCESS. dag_id=my_first_dag, task_id=multiply_my_number_by_23, execution_date=20220729T144715, start_date=20220729T151721, end_date=20220729T151722
```

> ## Step 9: Use the Astronomer Registry

Reviewer: @jwitz Let's talk about this, but I'm actually a little hesitant to add a step about the Astronomer Registry that isn't super defined. I'd rather either:

- Give users code they can copy paste (what we have above)
- Have users go find a specific DAG in the registry that we think is a good example

But I don't wanna say "step 9: go look at the registry" without more specific guidance. How do we know if they did it right or wrong? I just feel like we should pull out a specific DAG/use case.

I don't think "depending on what DAG you pick" is good and then trying to prevent errors without knowing exactly what the user is going to hit -- the registry is super intimidating, so we should guide experience more tangibly

Reviewer: @TJaniF What do you think? Would it be super hard to find a DAG from the registry to use in this example? I presume yes because 3rd party connections, but wondering if there's a way about that

Jake -- maybe we leverage same thing as Astro SDK?

Reviewer 3: @paolaperaza Yes I agree this step is daunting and not well defined. There is a registry DAG that works without 3rd party connections: https://registry.astronomer.io/dags/taskflow-etl . But then we of course can't show how connections work using that DAG. 

I was thinking maybe the user could add a task posting the result to Slack (assuming this is the tool most users have in common) to that DAG but I think the process of making a Slack app and creating a Webhook is a bit too much for a beginner and also either would take up a lot of space in the tutorial or the user would spend some time on Slack's documentation for it.

Another option would be to not show connections in this tutorial. Just mention them and link to the Managing connections guide (there is an update in PR state right now that will include a full example with screenshots there)? I think that could work with the structure:

Step 9: Copy a DAG from the Astronomer Registry
- Copy the Taskflow-ETL DAG
- Run it 
- Make a small change to the DAG that we explain in detail (maybe just add one more info from the API to the dictionary in task 2 and to the logging statement in task 3?)
- Run it again
- This DAG uses TaskFlowAPI syntax -> link out to guide
- A lot of registry DAGs use variables and connections -> link out to guide / docs

What do you think? 🙂

---
sidebar_label: 'Get started'

> title: 'Get started with Airflow using the Astro CLI'

Author: Is it Apache Airflow if we start the project with Runtime? What about just "Airflow"?

id: get-started

> description: Deploy your first local Airflow project with the Astro CLI.

Author: Im a little iffy because it's not _technically_ Apache Airflow but I guess the distinction doesn't really matter at this stage.

> description: Deploy your first local Airflow project with the Astro CLI.

Reviewer: It still is! I think that's ok, I'm not worried

---

Getting started with Airflow locally is made simple thanks to the Astro CLI.

This tutorial guides you through the process of using the Astro CLI to create a project, run your project in a local Airflow environment, and write DAGs. We will also cover how to use the Astronomer Registry to discover example DAGs and Airflow providers.

## Time to complete

This tutorial can be completed in 1 hour or less.

## Assumed knowledge

To get the most out of this tutorial, you should have knowledge of:

- Airflow fundamentals. See [Introduction to Apache Airflow](https://www.astronomer.io/guides/intro-to-airflow).
- Python fundamentals. See the [Python Documentation](https://docs.python.org/3/tutorial/index.html).
- How to provide arguments to Airflow operators. See [Operators 101](https://www.astronomer.io/guides/what-is-an-operator).

## Prerequisites

To use this tutorial, the following software is required:

- [Astro CLI](cli/configure-cli.md)
- [Docker Desktop](https://docs.docker.com/get-docker/) (v18.09 or higher)
- [Python 3](https://www.python.org/downloads/)

> - A terminal that accepts bash commands (pre-installed on most operating systems)

Author: The link here feels a bit gratuitous. Also "Terminal" is not a program (except on Mac). It's a kind of interface. Window's is called PoweShell

> - A terminal that accepts bash commands (pre-installed on most operating systems)

Reviewer: Fair -- let's not say `Terminal is pre-installed on most operating systems` then. Use whatever umbrella word, and make a statement about Windows and Mac?

Additionally, we recommend that you use an IDE for Python development, such as [VSCode](https://code.visualstudio.com/).

## Step 1: Create an Astro project

After [installing the Astro CLI](cli/configure-cli.md), create a folder for the new Astro project and run the initialization command from the terminal:

```sh
astro dev init
```

This command creates the folders and files necessary to run a Dockerized Airflow environment.

[Docker](https://docs.docker.com/) is a service to run software in virtualized containers within a machine. When you run Airflow locally with the Astro CLI, Docker creates these containers based the [Astro Runtime Docker image](https://docs.astronomer.io/astro/runtime-release-notes), which contains all necessary specifications on how to run the 4 [core Airflow components](https://www.astronomer.io/guides/airflow-components/). For this tutorial, no in-depth knowledge of Docker is needed. All you need to know is that Airflow will run in a containerized environment on your machine, and all necessary files for running these containers are included in the default Astro project.

The default Astro project includes the following files and folders:

> - `/.astro`: A set of configurations that should not be edited directly.

Author: Agreed but I'd also add the Dockerfile, just because it is the basis of how this all runs - plus we've already introduced Docker.

To access the Airflow UI, open `http://localhost:8080/` in a browser and log in with `admin` for both your username and password.

The default page in the Airflow UI is the **DAGs** page, which shows an overview of all DAGs in your Airflow environment:

![DAGView.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/ec24bc62-a4fb-4aa2-a1d0-fa0f9521d296/DAGView.png)

Each DAG is listed with its tags, owner, previous runs, schedule, timestamp of the last and next run, and the states of recent tasks. Because you haven't run any DAGs yet, the `Runs` and `Recent Tasks` sections are empty. Let's fix that!

## Step 4: Trigger a DAG run

1. Before you can run a DAG, you have to activate it. To activate a DAG, click on the slider button next to its name. Once activated, the DAG runs once for its latest scheduled date.

2. While all DAGs run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger an active DAG by clicking the play button under **Actions**. During development, manually triggering DAG runs can be helpful with debugging.

> ![DAGView_start_DAG.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/40421eb0-a4f6-4215-8562-d2040d97feea/DAGView_start_DAG.png)

Author: The actual information about instance success and colors feels a bit too advanced for this tutorial, but agreed we should explain what this result means and possibly include a screenshot

## Step 5: Explore the Airflow UI

> The Airflow UI's navigation bar contains 5 tabs:

Reviewer 4: Do we want to link out to our guide on the Airflow UI somewhere in here?

:::info

While DAG code can be viewed from within the Airflow UI, code edits have to be done directly in the Python file within the `/dags` folder. The displayed code will update when the scheduler parses the `/dags` folder, which by default happens every 30 seconds.

:::

## 6. Write a new DAG

Now that you know how to start and Airflow environment and run DAGs, you can start writing and testing your own DAGs.

1. Create a new Python file in the `/dags` folder named `my-dag.py`
2. Open the file in your IDE. Start by adding the required imports for Python packages:

    ```python
    from airflow import DAG
    from airflow.operators.python import PythonOperator
    from airflow.operators.bash import BashOperator

    from datetime import datetime, timedelta

    ```

    The first line imports the `DAG` class, the second and third line import different Airflow operators, and the last line imports two objects from the [datetime package](https://docs.python.org/3/library/datetime.html) which help define dates and time intervals for DAG scheduling.

> 3. Add two static variables and a simple Python function that multiplies an input by 23. These are can be used Airflow operators later in our DAG code.

Reviewer 4: ```suggestion
3. Add two static variables and a simple Python function that multiplies an input by 23. These are used by Airflow operators later in our DAG code.
```

Reviewer: I actually would add maaaybe half a sentence on what this DAG does -- "We will give this DAG an input from x and expect the multiple of 23 to show up in y" (bad phrasing but you know what I mean!)

    ```python
    # constants
    MY_NAME = "MY_NAME"
    MY_NUMBER = 19

    def multiply_by_23(number):
        """Multiplies a number by 23."""
        result = number * 23
        print(result)

    ```

4. Add a `dag` object, which is an instance of the `DAG` class:

    ```python
    with DAG(
        dag_id="my_first_dag",
        start_date=datetime(2022,7,28),
        schedule_interval=timedelta(minutes=30),
        catchup=False,
        tags= ["tutorial"],
        default_args={
            "owner": MY_NAME,
            "retries": 2,
            "retry_delay": timedelta(minutes=5)
        }
    ) as dag:

    ```

    `with DAG(...) as dag:` instantiates a DAG context in which Airflow operators can be defined and given depenencies. The instantiation includes several important arguments:

>     - `dag_id` (Required): The name of the DAG that appears in the Airflow UI. Each DAG must have unique name, and it's best practice to use the same name for both the DAG file and the `dag_id`.

Reviewer 4: I've never recommended this as a best practice before (and I often don't do it myself, although I probably should), but it would help keep the project organized and easier to read, and might help avoid accidentally duplicating dag_ids...I'm fine to leave it in, seems like a good thing for beginners to do :)

    - `schedule_interval`: How often the DAG should run. You can define this as a timedelta object, a [CRON expression](https://crontab.guru/), or a macro like "@daily". If you don't set this value, the DAG runs every 24 hours after the `start_date`.
    - `catchup`: Determines whether the DAG reruns all DAG runs that were scheduled before today's date. The default value is `True`, but it's best practice to set this argument to `False`.
    - `tags`: Defines the **Tags** that appear in the **DAGs** page of the Airflow UI.
    - `default_args`: A list of configurations for the DAG's behavior. In this DAG, we use these arguments to change the owner of the DAG, give the DAG a maximum of two retries in case it fails, and tell the DAG to wait five minutes before retrying. Many more arguments can be passed to a DAG at instantiation. See [Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dag/index.html) for a complete list.

4. Add an operator to your DAG:

    ```python
    t1 = BashOperator(
        task_id="say_my_name",
        bash_command=f"echo {MY_NAME}"
    )

    ```

>     Operators define the work to be completed by your DAG. They often encapsulate a single task or step of your DAG. Task 1 (`t1`) uses the [BashOperator](https://registry.astronomer.io/providers/apache-airflow/modules/bashoperator) to run a bash command printing the `MY_NAME` variable to the terminal. The `task_id` defines the name of the task that appears in the Airflow UI. Each task requires a unique `task_id`.

Reviewer 4: I think this sounds great!

>     We can now use `t1` to define dependencies within our DAG code.

Reviewer: Could explain what `t1` actually is? (Our second task (t2) below makes it clearer but should be higher)

Author: I think this line actually introduces more confusion than necessary because we don't define dependencies until later. 

As for explaining `t1`, I think the paragraph before does a pretty good job of explaining what t1 does. Explaining what t1 _is_ outside of "a task" seems like it might be out of the scope of this tutorial.

    This is one of the features that makes Airflow so powerful: Any action that can be defined in Python, no matter how complex, can be orchestrated using Airflow.

6. Define the dependencies between the two tasks using bit-shift operators:

    ```python
    t1 >> t2
    ```

    This notation is unique to Airflow. It both defines and visualizes the dependencies between tasks. `t1` is **upstream** of `t2`. meaning that `t1` must finish before `t2` can start.

    There are multiple ways to define dependencies in Airflow, but this is one of the simplest. For more options on how to define task dependencies, see [Managing Dependencies in Apache Airflow](https://www.astronomer.io/guides/managing-dependencies).

7. Save your code.

## Step 7: Run the new DAG

Refresh your browser at `http://localhost:8080/`. As long as Airflow is running, it automatically picks up any new changes in your `/dags` directory.

Click the name of your new DAG and open **Graph** view. The **Graph** view shows the dependencies between your tasks and is useful for troubleshooting issues that stem from dependencies.

> You can trigger a DAG run from any view in the **DAGs** page using the play button in the upper righthand corner of the UI. Trigger your new DAG from the **Graph** view.

Reviewer: This the same thing they did in a step above? Might be good to remind the user if so/use very consistent language

After your DAG runs, there should be a dark green border around the tasks showing that your run was successful.

![GraphView.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/95aadf3a-ded5-402f-84ff-bbe62099e0fb/GraphView.png)

> ### 8. Check out task logs

Reviewer: I'd probably just say "View task logs" for acccessibility

When we tell Airflow to print something to the terminal, the output appears in Airflow's logs. To see the output of our tasks, we need to open our tasks in either the **Grid** or the **Graph** view.  

> Click the `say_my_name` task to make the task instance view appear. From here, click the **Log** button.

Reviewer: Screenshot is prob necessary here

[2022-07-29, 15:17:19 UTC] {taskinstance.py:1415} INFO - Marking task as SUCCESS. dag_id=my_first_dag, task_id=say_my_name, execution_date=20220729T144715, start_date=20220729T151718, end_date=20220729T151719

```

Follow the same steps for the `multiply_my_number_by_23` task. The task logs should include the results of our multiplication operation.

```
[2022-07-29, 15:17:22 UTC] {logging_mixin.py:115} INFO - 437
[2022-07-29, 15:17:22 UTC] {python.py:173} INFO - Done. Returned value was: None
[2022-07-29, 15:17:22 UTC] {taskinstance.py:1415} INFO - Marking task as SUCCESS. dag_id=my_first_dag, task_id=multiply_my_number_by_23, execution_date=20220729T144715, start_date=20220729T151721, end_date=20220729T151722
```

## Step 9: Use the Astronomer Registry

The example DAG you wrote used two different core Airflow operators, but there are many more operators available outside of core Airflow. The best place to explore and search for operators is the [Astronomer Registry](https://registry.astronomer.io/).

The Astronomer Registry is also home to a growing set of [example DAGs](https://registry.astronomer.io/dags) that demonstrate more complex use cases and integrations.

> Pick an example DAG that uses tools you know and copy its code into a new file in your local `/dags` folder.

Reviewer: This seems almost silly easy, but might be good to have very explicit steps still?

- Go to Astronomer Registry
- Find a DAG
- Copy its contents (screenshot)
- Create a `.py` file in `dags` folder
- Paste DAG code there

Author: Agreed with this for the most part but I think the screenshot of the registry is unnecessary

If you are new to the world of data engineering and haven't used other tools yet, you can use the [TaskFlow API ETL Example](https://registry.astronomer.io/dags/taskflow-etl) which demonstrates how you can define task dependencies using the TaskFlow API.

> Depending on which DAG you picked, Airflow might produce an import error. This is because the DAG uses modules that are not part of core Airflow.

Reviewer: Could maybe explain this a tad bit more

![ImportError.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/9da60514-66c0-4500-9987-9a6bc62bc6a3/ImportError.png)

These modules have to be made available to your Docker image before you run the `astro dev start` command. Search the registry for the missing package, click on the provider, and copy its name.

![SearchProvider.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/d97c954d-3b50-4b0c-a7d8-b7422284c0b3/SearchProvider.png)

Paste this name on its own line in the `requirements.txt` file of your Astro project. If you want, you can also install a specific version like in the following example:

> ```

Reviewer: Hm not sure exactly what we need to add here, but seems like we may be losing the user if they don't know what they're doing

Author: I tried making it easier to follow by cutting this out into its own section and providing steps

> Paste this name on its own line in the `requirements.txt` file of your Astro project. If you want, you can also install a specific version like in the following example:
> 
> ```
> apache-airflow-providers-common-sql==1.0.0
> ```
> 

Reviewer 3: ```suggestion
Paste this name on its own line in the `requirements.txt` file of your Astro project. You can add as many Python packages as you need. If you want, you can also install a specific version like of the Common SQL provider in the following example:

```
apache-airflow-providers-common-sql==1.0.0
apache-airflow-providers-amazon
scikit-learn
```

If you have tasks in your DAG needing different versions of the same package or dependency conflicts consider using the [KubernetesPodOperator](https://www.astronomer.io/guides/kubepod-operator/) to run a task in a separate environment.

```

Adding this as a possibility as discussed in the slack channel @kentdanas @paolaperaza . Not sure if it is a bit much for a beginner hearing about KPO but it could also be helpful for experienced DE people who are simply new to Airflow. 🤔

Reviewer 4: Hmm, I'm on the fence about whether to mention KPO, would be curious to hear other's opinions. I think KPO is a bit advanced for to mention in this tutorial, but I also think it would be weird if we mentioned potential dependency conflicts and didn't provide some link out to a fix....

Reviewer: I think KPO mention is fine? May honestly just keep it as an explicit "tip" -- like, recmmend the thing in the context of a future problem vs throw it out as something they might want to do during this tutorial. Presumably we don't want them touching KPO in this tutorial, otherwise we'd have steps for it

Author: I'd opt for not including it, because we're kind of skirting the issue here and making this seem more simple than it is. "Run the KPO" requires running an entire Kubernetes namespace, which I'm guessing most people following this tutorial won't have access to.

> If you make changes to `requirements.txt`, `packages.txt`, or `Dockerfile`, or you add new files to `/include` or `/plugins`, you need to restart your Airflow environment to build your changes.

Reviewer 4: ```suggestion
If you make changes to `requirements.txt`, `packages.txt`, or `Dockerfile`, you need to restart your Airflow environment to build your changes.
```

include/ and plugins/ are mounted directories, so changes within them are picked up automatically without a restart.

You can restart Airflow using the Astro CLI by running:

```sh
astro dev restart
```

## Step 10: Add variables and connections

Some DAGs require additional configurations like setting Airflow [variables](https://airflow.apache.org/docs/apache-airflow/stable/howto/variable.html) and [connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html). Airflow variables define values for your DAGs outside of your DAG code, while connections define how Airflow integrates with external systems. Take the following code, for example:

> ```python

Reviewer: Should the user be adding this to their project?

Author: This is just a snippet. Tried to clarify that

    was uploaded without errors.
    """
    s3 = S3Hook()
    # This code requires an Airflow variable that contains a JSON file of AWS configurations
    aws_configs = Variable.get("aws_configs", deserialize_json=True)
    obj = s3.get_key(
        key=f"{aws_configs.get('s3_key_prefix')}/{CSV_FILE_PATH}",
        bucket_name=aws_configs.get("s3_bucket"),
    )
    obj_etag = obj.e_tag.strip('"')
    file_hash = hashlib.md5(
        open(CSV_FILE_PATH).read().encode("utf-8")).hexdigest()
    if obj_etag != file_hash:
        raise AirflowException(
            f"Upload Error: Object ETag in S3 did not match hash of local file."
        )
```

If the code contains `Variable.get("variable_name")`, you need to add a variable of that name with a fitting input. If the code contains a `conn_id` parameter, you need to define an Airflow connection.

To create connections, go to **Admin** > **Connections** in the Airflow UI. In the case of the code above, we need to define an `aws_default` connection. For more information about managing connections, see [Managing your Connections in Apache Airflow](https://www.astronomer.io/guides/connections/).

> Airflow variables can be defined in **Admin** > **Variables**. However, it's best practice to define them as environment variables in your Astro project's `.env` file. All custom environment variables used by Airflow must start with `AIRFLOW_VAR`:

Reviewer 3: Good idea to add that, thanks!

```text
AIRFLOW_VAR_MY_VAR="my-value"
AIRFLOW_VAR_AWS_CONFIGS="<your-config-json>"
```

Environment variables can be accessed from within the DAG alongside your imports:

```python
import os
my_var = os.environ["AIRFLOW_VAR_AWS_CONFIGS"]
```

> ### 9.4 TaskFlow API

Reviewer 4: If we do keep this section, let's link to the decorators guide somewhere in it.

Also, it seems a bit odd to me to have a section that doesn't have an action for the users to complete... I would lean towards just having a note about @task syntax in Step 9 because this feels a bit out of place coming after Step 10.

> ### 9.4 TaskFlow API

Reviewer: Tend to agree with @kentdanas (to leave it out) but I think could mention it as a "tip"

> ### 9.4 TaskFlow API

Author: I agree that we can delete this section. If possible, I'd also like to rewrite the example code in the connections/ variables step to use the more basic syntax, just so users can focus on the task at hand.

        print(result)

    t1 >> multiply_by_23()

dag = my_first_dag_2()

```

Note the last line of the code example: DAGs defined using the `@dag()` decorator must be called to appear in Airflow.

Whether to use regular operators or decorated tasks is often a question of personal preference. The TaskFlow API offers more abstraction and generally requires less code.

## Conclusion

After completing this tutorial, you should know how to:

- Run a local Airflow environment using the Astro CLI.
- Manage the Astro project directory.
- Write a DAG from scratch.
- Navigate the Airflow UI.
- Use code from the Astronomer registry.

### Next steps

Astronomer offers a variety of resources like this tutorial to learn more about how to use Airflow.

- [Astronomer Webinars](https://www.astronomer.io/events/webinars/) cover concepts and use cases in-depth and offer the possibility to ask us questions live on air.
- [Live with Astronomer](https://www.astronomer.io/events/live/) are hands-on and code focussed live walkthroughs of specific Airflow features.
- [Astronomer Guides](https://www.astronomer.io/guides/) cover both entry and expert level concepts in Airflow.

> - [Astronomer Academy](https://academy.astronomer.io/) offers many video tutorials and the option to purchase full length Airflow courses and to take exams to get certified.

Reviewer 4: Is the next 'Don't know where to start?' section not enough for directing users to second step resources? Seems like we should have one or the other.

> - [Astronomer Academy](https://academy.astronomer.io/) offers many video tutorials and the option to purchase full length Airflow courses and to take exams to get certified.

Reviewer 3: Yes, both sections would be a bit much. The idea of my edit was to roll them into one section. Have the link to the general resources but also point users to a specific piece for each to go next. 
But I think it could also work without linking to the general resource, just by having each bullet point be in the style of:

"Astronomer Webinars cover concepts and use cases in-depth and the possibility to ask us questions live on air. Check out this Webinar on [Writing Functional DAGs with Decorators](https://www.astronomer.io/events/webinars/writing-functional-dags-with-decorators/)."

> Don't know where to start? For beginners, the next resources we recommend are:

Reviewer 3: ```suggestion
```
